src/plugins/kibana/public/settings/sections/indices/add_data_steps/pipeline_setup/directives/pipeline_setup.js

```diff
@@ -0,0 +1,87 @@
+import uiModules from 'ui/modules';
+import _ from 'lodash';
+import Pipeline from '../lib/pipeline';
+import Processor from '../lib/processor';
+import angular from 'angular';
+import processorTypes from '../lib/processor_types';
+import savedPipeline from '../sample_pipeline.json'; //temp for debugging purposes
+
+require('../services/ingest');
+require('../styles/_pipeline_setup.less');
+require('./pipeline_output');
+require('./source_data');
+require('./processors');
+
+const app = uiModules.get('kibana');
+
+app.directive('pipelineSetup', function () {
+  return {
+    restrict: 'E',
+    template: require('../views/pipeline_setup.html'),
+    scope: {
+      samples: '=',
+      pipeline: '='
+    },
+    controller: function ($scope, ingest, debounce, Notifier) {
+      const notify = new Notifier({ location: `Ingest Pipeline Setup` });
+      $scope.processorTypes = _.sortBy(processorTypes, 'title');
+      $scope.sample = {};
+
+      const pipeline = new Pipeline(processorTypes, Processor);
+      if ($scope.pipeline) {
+        pipeline.load($scope.pipeline);
+        $scope.sample = $scope.pipeline.input;
+      }
+      $scope.pipeline = pipeline;
+
+      //initiates the simulate call if the pipeline is dirty
+      function simulatePipeline(event, message) {
+        if (!pipeline.dirty) return;
+
+        if (pipeline.processors.length === 0) {
+          pipeline.updateOutput();
+          return;
+        }
+
+        ingest.simulatePipeline(pipeline)
```

Comment:
Best practice. It makes your intentions explicit and if someone comes along and adds code after the last line, static analysis will tell them their code is unreachable instead of silently changing the return value of the function.